### PR TITLE
Make settings files per-world.

### LINKED ISSUE #6966: World-specific settings
##### Issue type
 - Feature request
 
##### Summary
Currently, most settings get mangled together in a single global `minetest.conf`. But you quickly run into problems when there's a setting for which different values for different worlds make sense, rather than a globally shared single value.

The most striking example is `time_speed`. I have to always manually change this between 0 and 72 when I switch between development worlds (where I don't want the day/night cycle) and playing worlds (where I want it).

Examples for candidates for world-speficic settings:
* `time_speed`
* `enable_pvp`
* `enable_damage`
* Mod stuff: Disable mob spawning, disable weather changes, disable fire, disable TNT, …

I suggest to add a new setting category: “World-specific setting”. This is a setting which is set on a per-world basis. This would allow me to have a different `time_speed` for each world and I never have to bother about it again.

Minetest already implements this to some extent, but there are holes in the system. There is no standard (!) way to add a custom world-specific setting. There is no server command either.

Deprecates https://github.com/minetest/minetest/issues/5353.

### PR DESCRIPTION
**Pull request by red-001**
### Core changes

Settings are now loaded from `world.mt` on server start up, with `minetest.conf` settings being used as defaults.

### Main menu Changes

The biggest change is a new menu based on the advanced settings menu that can be used to set per-world settings. A few small changes were also made to the "start game" tab to better support then new way that `world.mt` is used.

This PR replaces #6528 and is a simpler implementation, it still allows for separating client and server settings, WIP as it needs testing and some input on which client settings should be always global. It also closes #6966.

### CODE REVIEW

**Comment by rubenwardy** [comment]
Nice, much simpler than the previous implementation

**Comment by paramat** [comment]
Testing:

Clicking on 'world settings' with no worlds present causes an error.

I created world X and entered 'world settings', this still has 'return to settings page' as a button (wrong label).
I made 'water level' 100 and saved this.
I entered 'advanced settings' (which are now default world settings) and set 'water level' to -100.
Started world X, water level is normal.
Looking in the world folder there is both world.mt and map_meta.txt. world.mt has 'water_level = 100' and map_meta.txt has 'water_level = 1'
Started a second world Y to use the default settings in advanced settings menu, water level is -100 as it should be.
World Y also has both world.mt and map_meta.txt, in map_meta txt ther is 'water_level = -100' which is correct.

I'll just repeat this to confirm.

**Comment by paramat** [comment]
IRC discussion http://irc.minetest.net/minetest-dev/2018-02-13#i_5230374
Explains many of the bugs in testing.

22:19 red-001     well thought I would ask you since I'm considering also merging map_meta.txt with world.mt, and mapgen is kinda your field
22:35 paramat     ah hmm thanks
22:36 paramat     oh it's more radical than i realised, will look over it
22:38 Fixer       this one seems good for review, not too big https://github.com/minetest/minetest/pull/6825
22:38 paramat     what is the reason for merging those 2 files?
22:39 red-001     most of the C++ changes are just to avoid client settings like view range being saved in the world config
22:39 red-001     paramat, to make experimenting with mapgen flags easier
22:39 Fixer       and this one too needs reviews, important https://github.com/minetest/minetest/pull/6786
22:41 red-001     also makes it easier for someone to mess up thier world which is why I haven't implemented it
22:43 paramat     will test 6982
22:45 paramat     6982 is big so needs lots of attention
22:47 red-001     Fixer, is 6825 only for cinematic?
22:48 red-001     or does cimematic just happen to be enabled?
22:48 Fixer       red-001: no, smooth yaw for objects, for example, carts, mobs, etc
22:51 paramat     so i create a world, highlight it, then click 'world settings'?
22:52 red-001     that should open a modified version of advance settings with anything thats client only removed
22:53 paramat     ok, it's very buggy, will reply in thread
23:14 red-001     paramat, like I said I haven't merged map_meta.txt with settings, just was considering doing so
23:17 paramat     ok
23:17 paramat     oh yeah misread
23:18 paramat     is this why altering mapgen parameters doesn't work?
23:18 red-001     yes
23:18 paramat     ok
23:18 red-001     fixed the two other issues
23:20 paramat     map meta txt is already per world so setting mapgen parameters in world settings may be irrelevant, of course, duh
23:21 paramat     but then you'd need to disallow changing some mapgen params in world creation i guess if they aren't going to work
23:22 red-001     well since mapgen now seems to be it's own top-level settings group I could just filter it out easily
23:23 paramat     i think at the moment i prefer map meta txt stay separate and contain stuff that is already per-world, and therefore not be settable in world settings
23:23 red-001     done
23:23 paramat     mapgen params would then be set in advanced settings as before

**Comment by paramat** [comment]
I see you have removed mapgen settings from the per-world settings menu but how do the settings APIs work now? I assume settings.get/set act on the per-world settings? But how do the mapgen setting APIs act?

**Comment by Wuzzy2** [comment]
Screenshots?

**Comment by red-001** [comment]
@Wuzzy2 
screenshot from 2018-02-25 13-09-07
screenshot from 2018-02-25 13-12-15
screenshot from 2018-02-25 13-11-45

**Comment by red-001** [comment]
@paramat the PR is fairly simple, most of the C++ changes are either there to make sure that settings that should be global (e.g. keymappings) stay that way or to avoid having other code directly read or write to `world.mt` rewriting it to use `g_settings` instead. The part that settings up `g_settings` is pretty short and fairly self-explanatory, see https://github.com/minetest/minetest/pull/6982/files#diff-ad60d65b34e16a3319296bb5d683acd6R171

File: builtin/mainmenu/common.lua [hunk 1]
```diff
@@ -303,34 +303,31 @@ function is_server_protocol_compat_or_error(server_proto_min, server_proto_max)
...
-		else
-			menu_worldmt(selected, mode_name, core.settings:get(mode_name))
-		end
+function get_world_setting(selected, setting)
+	local world_conf = get_world_config(selected)
+	if not world_conf then
+		world_conf = core.settings
+	end
+
+	return world_conf:get(setting) or core.settings:get(setting)
```
Comment (SmallJoker) [inline builtin/mainmenu/common.lua]: Simpler: `return (world_conf and world_conf:get(setting)) or core.settings:get(setting)`

File: builtin/mainmenu/common.lua [hunk 2]
```diff
@@ -303,34 +303,31 @@ function is_server_protocol_compat_or_error(server_proto_min, server_proto_max)
...
+end
+
+function set_world_setting(selected, setting, value)
+	local world_conf = get_world_config(selected)
+	if not world_conf or not value then return end
```
Comment (SmallJoker) [inline builtin/mainmenu/common.lua]: Please use multiple lines.

File: builtin/mainmenu/dlg_settings_advanced.lua [hunk 3]
```diff
@@ -854,7 +859,28 @@ local function handle_change_setting_buttons(this, fields)
...
-local function create_settings_formspec(tabview, name, tabdata)
+local function filter_world_settings()
+	local current_level = 0
+	local settings = {}
+	local ignore_top_level_category = false
+	local world_settings_categories = {"Server / Singleplayer", "Client and Server", "Games", "Mods"}
+	for _, entry in ipairs(full_settings) do
+		if entry.type == "category" and entry.level == 0 then
+			ignore_top_level_category = true
+			for _, allowed_catergory in ipairs(world_settings_categories) do
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_settings_advanced.lua]: Typo: `allowed_category`

File: builtin/mainmenu/dlg_settings_advanced.lua [hunk 4]
```diff
@@ -854,7 +859,28 @@ local function handle_change_setting_buttons(this, fields)
 	return false
 end
 
-local function create_settings_formspec(tabview, name, tabdata)
+local function filter_world_settings()
+	local current_level = 0
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_settings_advanced.lua]: Unused variable

File: builtin/mainmenu/dlg_settings_advanced.lua [hunk 5]
```diff
@@ -979,11 +1018,11 @@ local function handle_settings_buttons(this, fields, tabname, tabdata)
 		if setting and setting.type ~= "category" then
 			if setting.type == "noise_params_2d"
 					or setting.type == "noise_params_3d" then
-				core.settings:set_np_group(setting.name, setting.default_table)
+			conf:set(setting.name, setting.default)
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_settings_advanced.lua]: Missing indent. Why not `set_np_group`?

File: builtin/mainmenu/tab_local.lua [hunk 6]
```diff
@@ -280,6 +274,21 @@ local function main_button_handler(this, fields, name, tabdata)
 
 		return true
 	end
+	
+	if fields["world_settings"] ~= nil then
```
Comment (SmallJoker) [inline builtin/mainmenu/tab_local.lua]: Superfluous `~= nil`. It's never `false`, i.e. always either a string or nil.

File: src/main.cpp [hunk 7]
```diff
@@ -893,13 +900,16 @@ static bool run_dedicated_server(const GameParams &game_params, const Settings &
...
 	Settings world_mt;
+	set_default_settings(&world_mt);
+	override_default_settings(&world_mt, g_main_settings);
 	std::string world_mt_path = game_params.world_path + DIR_DELIM + "world.mt";
+
 	if (!world_mt.readConfigFile(world_mt_path.c_str())) {
 		errorstream << "Cannot read world.mt!" << std::endl;
 		return false;
 	}
+	g_settings = &world_mt;
```
Comment (SmallJoker) [inline src/main.cpp]: Pointer assigned to address of local variable. If you want to keep this, please ensure `g_setting` points to valid data when the function returns.

File: src/server.cpp [hunk 8]
```diff
@@ -155,24 +155,36 @@ Server::Server(
...
 	m_uptime(0),
-	m_clients(m_con),
 	m_admin_chat(iface),
 	m_modchannel_mgr(new ModChannelMgr())
 {
+	// Initialize settings
+	set_default_settings(m_conf);
+	override_default_settings(m_conf, g_main_settings);
+	m_settings_path = path_world + DIR_DELIM + "world.mt";
+	if(!m_conf->readConfigFile(m_settings_path.c_str()))
```
Comment (SmallJoker) [inline src/server.cpp]: Missing space after `if`

File: src/terminal_chat_console.cpp [hunk 9]
```diff
@@ -347,10 +347,7 @@ void TerminalChatConsole::step(int ch)
 			continue;
 
 		std::wstring error_message = utf8_to_wide(Logger::getLevelLabel(p.first));
-		if (!g_settings->getBool("disable_escape_sequences")) {
-			error_message = std::wstring(L"\x1b(c@red)").append(error_message)
-				.append(L"\x1b(c@white)");
-		}
```
Comment (SmallJoker) [inline src/terminal_chat_console.cpp]: Is this change intended?

Comment (red-001) [reply src/terminal_chat_console.cpp]: no sorry that was meant for/should be in another PR.

File: builtin/mainmenu/dlg_settings_advanced.lua (hunk 5 above)
Comment (red-001) [reply builtin/mainmenu/dlg_settings_advanced.lua]: I think that was just an oversight.

File: src/content/subgames.cpp [hunk 10]
```diff
@@ -37,6 +37,21 @@ bool getGameMinetestConfig(const std::string &game_path, Settings &conf)
...
+{
+	std::string conf_path = game_path + DIR_DELIM + "game.conf";
+	return conf.readConfigFile(conf_path.c_str());
+}
+
+std::string getGameName(const std::string &game_path)
+{
+	Settings conf;
+	if(!getGameConfig(game_path, conf) && !conf.exists("name"))
+		return "";
```
Comment (rubenwardy) [inline src/content/subgames.cpp]: this fails if there is no conf and name in conf, whereas the rest of the code fallbacks to the folder name

File: src/content/subgames.cpp [hunk 11]
```diff
@@ -283,35 +298,36 @@ std::vector<WorldSpec> getAvailableWorlds()
...
-	getGameMinetestConfig(gamespec.path, game_defaults);
-	override_default_settings(g_settings, &game_defaults);
-
 	infostream << "Initializing world at " << path << std::endl;
 
 	fs::CreateAllDirs(path);
 
+	Settings conf;
+	set_default_settings(&conf);
+	override_default_settings(&conf, g_settings);
```
Comment (rubenwardy) [inline src/content/subgames.cpp]: was this meant to change order?

File: src/content/subgames.cpp [hunk 12]
```diff
@@ -283,35 +298,36 @@ std::vector<WorldSpec> getAvailableWorlds()
...
 	if (!fs::PathExists(worldmt_path)) {
-		Settings conf;
-
 		conf.set("gameid", gamespec.id);
-		conf.set("backend", "sqlite3");
-		conf.set("player_backend", "sqlite3");
-		conf.setBool("creative_mode", g_settings->getBool("creative_mode"));
-		conf.setBool("enable_damage", g_settings->getBool("enable_damage"));
+
+		SETUP_SETTING(backend);
```
Comment (rubenwardy) [inline src/content/subgames.cpp]: if the key isn't in conf, what value will this take?

File: src/content/subgames.cpp [hunk 13]
```diff
@@ -283,35 +298,36 @@ std::vector<WorldSpec> getAvailableWorlds()
...
+		SETUP_SETTING(player_backend);
+		SETUP_SETTING(creative_mode);
+		SETUP_SETTING(enable_damage);
 
 		if (!conf.updateConfigFile(worldmt_path.c_str()))
 			return false;
+	} else {
+		// read settings for use by map_meta.txt
+		conf.readConfigFile(worldmt_path.c_str());
```
Comment (rubenwardy) [inline src/content/subgames.cpp]: conf isn't stored into g_settings after here

File: src/main.cpp (hunk 7 above)
Comment (rubenwardy) [reply src/main.cpp]: :O

File: src/content/subgames.cpp (hunk 12 above)
Comment (red-001) [reply src/content/subgames.cpp]: default

File: builtin/mainmenu/dlg_settings_advanced.lua [hunk 14]
```diff
@@ -854,7 +859,27 @@ local function handle_change_setting_buttons(this, fields)
 	return false
 end
 
-local function create_settings_formspec(tabview, name, tabdata)
+local function filter_world_settings()
```
[the remark this answers is not in the document]
Comment (red-001) [reply builtin/mainmenu/dlg_settings_advanced.lua]: Currently it only filters top-level catergories, aren't mod settings are all under the mods catergory? You are right it would be better to add this as an attribute to the settingtypes file.

File: src/content/subgames.cpp (hunk 11 above)
Comment (red-001) [reply src/content/subgames.cpp]: yes, it initializes the default settings and then overrides them with the values from `minetest.conf`

**Comment by paramat** [comment]
Seems neglected with requests unattended to. Will reopen on progress.

**Comment by kneekoo** [comment]
Can someone please pick this back up? Luanti matured a lot lately, and isolating the world settings would be great, especially `time_speed`.

**Comment by sfan5** [comment]
There is an open issue for this (#6966), please don't bump ancient PRs.

> especially time_speed.

what's the problem with time_speed?

**Comment by kneekoo** [comment]
Given the last comment by paramat, and the fact that it was the most recent PR attempting to implement world-specific settings, it made sense to draw attention here.

Players unaware of games and mods using `time_speed` will exit that world and play something else, and the way everything would work becomes unpredictable just because `time_speed` is a global setting. It can be a problem even when multiple worlds run on the same game "compete" by changing this setting to match what's going on there.

State can be saved to prevent weird stuff, but games not using this setting will obviously be affected. Even the mods relying on the default `time_speed` value will change their behavior in unwanted ways.

**Comment by sfan5** [comment]
Obviously games messing with global settings will mess things up, but as far as I'm aware there is *no reason* for games to set `time_speed` globally. In fact I'd consider it a bug if a game does that.

> Even the mods relying on the default time_speed value will change their behavior in unwanted ways.

If your mod relies on `time_speed` being the engine's default value it is also broken by design.

**Comment by kneekoo** [comment]
As far as I can tell from reading here and in other issues, that setting is strictly global. There are reasons for games and mods to [change it in specific worlds](https://github.com/luanti-org/luanti/issues/14922), but of course there are other reasons to want to be able to change this per-world, since imagination can go pretty far.

I only asked because I see there are others who want this done and I was hoping that once it's done it becomes an option for new game content/features.
